fix: Ensure DaemonSets with '/initialized' and '/registered' are considered during nodeclaim calculations

1mwataru opened the pull request:
Fixes #2116 

**Description**

- When Daemonsets has nodeAffinity or nodeSelector related to `karpenter.sh/registered` and `karpenter.sh/initialized` labels, these are not considered during NodeClaim calculations, resulting in nodes being launched with smaller sizes than expected.
  - This leads to unexpected behaviors such as resource shortages on nodes, preventing Daemonset Pods running, or Deployment Pods remaining in Pending state causing new NodeClaims to be created.

- Modified the code to consider `karpenter.sh/registered` and `karpenter.sh/initialized` labels during Daemonset overhead calculations.

**How was this change tested?**

- Prerequisites
  - Prepared the following Daemonset and Deployment
    - Daemonset
      - Pod Identity
        - No setting
      - VPC CNI (aws-node)
        - Cpu request: 25m
      - Kube Proxy
        - Cpu request: 100m
      - My Daemonset (with required nodeAffinity that specifies nodeSelectorTerms with matchExpressions for the `karpenter.sh/registered` label)
        - Cpu request: 2 (2000m)
    - Deployment
      - My Deployment
        - Cpu request: 1 (1000m)

- Before the change
  - Confirmed from Karpenter logs that My Daemonset was not considered in the request value (`"cpu":"1150m"`)
   
    ```
    {"level":"INFO","time":"2025-04-23T13:11:26.806Z","logger":"controller","caller":"provisioning/provisioner.go:413","message":"created nodeclaim","commit":"3f9eb00-dirty","controller":"provisioner","namespace":"","name":"","reconcileID":"e786ef2c-95ea-4607-b317-d684dd519b89","NodePool":{"name":"my-nodepool"},"NodeClaim":{"name":"my-nodepool-jtzvc"},"requests":{"cpu":"1150m","pods":"4"},"instance-types":"c5.large, c5.xlarge, c5a.large, c5a.xlarge, c5d.large and 55 other(s)"}
    ```

  - As a result, an EC2 instance of type c5d.large (allocatable cpu: 1930m) was launched, but the Daemonset remained in Pending state

- After the change
  - Confirmed from Karpenter logs that My Daemonset was now considered in the request value (`"cpu":"3150m"`)

    ```
    {"level":"INFO","time":"2025-04-23T13:45:02.609Z","logger":"controller","caller":"provisioning/provisioner.go:413","message":"created nodeclaim","commit":"3f9eb00-dirty","controller":"provisioner","namespace":"","name":"","reconcileID":"061608ab-f9ad-41fa-ade5-fd19a2ccec58","NodePool":{"name":"my-nodepool"},"NodeClaim":{"name":"my-nodepool-wbh4m"},"requests":{"cpu":"3150m","pods":"5"},"instance-types":"c5.2xlarge, c5.xlarge, c5a.2xlarge, c5a.xlarge, c5d.xlarge and 55 other(s)"}
    ```

  - The Daemonset started successfully without issues
 
**Manifests used for verification**

- NodePool
  
  <details><summary>nodepool.yaml</summary>
  <p>

    ```yaml
    apiVersion: karpenter.sh/v1
    kind: NodePool
    metadata:
    name: my-nodepool
    spec:
    template:
        metadata:
        labels:
            cluster-name: cluster-31-3923-alb
        spec:
        nodeClassRef:
            group: karpenter.k8s.aws
            kind: EC2NodeClass
            name: default
        requirements:
        - key: karpenter.k8s.aws/instance-category
            operator: In
            values:
            - m
            - c
            - r
        - key: karpenter.k8s.aws/instance-generation
            operator: Gt
            values:
            - "4"
        - key: kubernetes.io/arch
            operator: In
            values:
            - amd64
        - key: kubernetes.io/os
            operator: In
            values:
            - linux
    ```

  </p>
  </details> 

- Deployment and Deamonset

  </p>
  </details> 
  <details><summary>dep_deamon.yaml</summary>

    ```yaml
    apiVersion: apps/v1
    kind: Deployment
    metadata:
    name: my-deploy
    spec:
    replicas: 1
    selector:
        matchLabels:
        jeremy: my-deploy
    template:
        metadata:
        labels:
            jeremy: my-deploy
        spec:
        nodeSelector:
            karpenter.sh/nodepool: my-nodepool
        containers:
        - name: my-dep-container
            image: nginx:latest
            ports:
            - containerPort: 80
            resources:
            requests:
                cpu: "1"

    ---

    apiVersion: apps/v1
    kind: DaemonSet
    metadata:
    name: my-daemonset
    spec:
    selector:
        matchLabels:
        app.kubernetes.io/name: my-daemonset-app
    template:
        metadata:
        labels:
            app.kubernetes.io/name: my-daemonset-app
        spec:
        affinity:
            nodeAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
                nodeSelectorTerms:
                - matchExpressions:
                - key: karpenter.sh/registered
                    operator: In
                    values:
                    - "true"
                - key: karpenter.sh/nodepool
                    operator: In
                    values:
                    - my-nodepool
        containers:
        - image: nginx:latest
            imagePullPolicy: IfNotPresent
            name: my-ds-container
            resources:
            requests:
                cpu: "2"
        tolerations:
        - operator: Exists
    ```

  <p>

jonathan-innis commented on pkg/apis/v1/labels.go:
```diff
@@ -89,6 +89,14 @@ var (
+	// AllowedDaemonsetLabels is a set of label keys permitted for Daemonset resources.
+	// For Daemonsets, Karpenter also considers labels indicating registration and initialization status
+	// when performing scheduling.
+	AllowedDaemonsetLabels = sets.New(
```
This isn't quite right, right? Because we know that the value is going to be set to `true`, I don't think we can make the change within the requirements checking code. 

It's probably more fitting to do this in the same place we inject the `karpenter.sh/nodepool` label: https://github.com/kubernetes-sigs/karpenter/blob/main/pkg/controllers/provisioning/scheduling/nodeclaimtemplate.go#L68.

We do have to be careful that when we actually create the NodeClaim, we don't add these labels on there then, but other than that, I think that would fix the scheduling representation that we have

1mwataru replied:
Thank you for your comment!
I have modified it to add within the nodeclaimtemplate.

To minimize changes to the original NodeClaimTemplate Requirements, I am not using `delete()` when removing items during NodeClaim creation.
Please let me know if using `delete()` would be better from a performance or memory perspective.

jonathan-innis commented:
/assign @jonathan-innis

1mwataru committed:
fix: Ensure DaemonSets with '/initialized' and '/registered' are considered during nodeclaim calculations

jonathan-innis commented on pkg/controllers/provisioning/scheduling/nodeclaimtemplate.go:
```diff
@@ -70,6 +70,12 @@ func NewNodeClaimTemplate(nodePool *v1.NodePool) *NodeClaimTemplate {
 	nct.Requirements.Add(scheduling.NewNodeSelectorRequirementsWithMinValues(nct.Spec.Requirements...).Values()...)
 	nct.Requirements.Add(scheduling.NewLabelRequirements(nct.Labels).Values()...)
+
+	// Add requirements for DaemonSet scheduling calculations
```
This looks good! Can we add one test to make sure that this works properly on the two new labels? Maybe something with affinity?

jonathan-innis commented on pkg/controllers/provisioning/scheduling/nodeclaimtemplate.go:
```diff
@@ -80,6 +86,14 @@ func (i *NodeClaimTemplate) ToNodeClaim() *v1.NodeClaim {
 		return i.Name
 	})...))
 
+	// Filter out DaemonSet scheduling-only requirements for the actual NodeClaim
```
nit: These may not just be DaemonSet scheduling requirements -- I'm not sure why you would do this on things other than DaemonSets, but we should technically support this for any pod that selected against this label

1mwataru replied:
@jonathan-innis 
Thanks for your comment.

Apologies for the delay.
We exclude other workloads like Deployments due to the following scenarios:

- When node affinity for these labels are applied to Deployments, current [validation](https://github.com/kubernetes-sigs/karpenter/blob/main/pkg/apis/v1/labels.go#L127) rejects them during NodeClaim creation, preventing NodeClaim creation entirely
- Simply removing this validation leads to unintended behavior:
  - Deployments create Pending Pods, triggering new NodeClaim creation
  - Pods remain unscheduled until Nodes become registered/initialized, causing redundant NodeClaim creation

Addressing this would require logic to prevent excessive NodeClaim creation when Pending Pods exist with registered/initialized labels. 
This represents a not small change beyond this Issue's scope, so we excluded it.
(Please let me know if we should create a separate Issue for investigation, or if any comments need modification.)

1mwataru committed:
add test

1mwataru committed:
refactor: replace ForEachNode with range loop in scheduling tests

1mwataru commented:
/remove-lifecycle stale

Hi, @engedaam @jonathan-innis @tallaxes

I've confirmed that this issue is still not fixed in the latest release.

Would you mind reviewing it when you get a chance?

Thanks!

1mwataru committed:
refactor: simplify requirements filtering and clarify comments

jmdeal commented:
/assign @jmdeal
/unassign @jonathan-innis

1mwataru committed:
fix: prevent registered/initialized requirements from leaking into NodeClaim labels and consolidate exclusion logic

Signed-off-by: Wataru Imai <imwata@amazon.com>

1mwataru commented on pkg/controllers/provisioning/scheduling/nodeclaimtemplate.go:
```diff
@@ -81,7 +96,7 @@ func NewNodeClaimTemplate(nodePool *v1.NodePool) *NodeClaimTemplate {
 	labels := map[string]string{}
 	for key, requirement := range i.Requirements {
-		if v1.WellKnownLabels.Has(key) || v1.RestrictedLabels.Has(key) {
+		if v1.WellKnownLabels.Has(key) || v1.RestrictedLabels.Has(key) || schedulingSimulationKeys.Has(key) {
```
Due to this method and my previous changes, the `karpenter.sh/initialized=true` label was being assigned at NodeClaim creation time as well. 
This change excludes the registered/initialized labels as a special case to prevent that.

1mwataru commented:
Hi @jmdeal ,

Thank you for stepping in as the reviewer.

I've resolved the conflicts. Could you please take another look when you get a chance?

Thanks!

AndrewMitchell25 reviewed (commented):
Thanks for this fix — the approach is solid and the DaemonSet overhead calculation fix is well-validated.

One suggestion: could you add assertions verifying that the simulation-only requirements (`karpenter.sh/registered` and `karpenter.sh/initialized`) don't leak into the created NodeClaim? This would validate the `ToNodeClaim()` stripping logic alongside the overhead calculation.

Right after `node1 := ExpectScheduled(ctx, env.Client, initialPod)`, something like:

```go
// Verify simulation-only requirements don't leak into the created NodeClaim
Expect(cloudProvider.CreateCalls).To(HaveLen(1))
createdNodeClaim := cloudProvider.CreateCalls[0]
reqs := pscheduling.NewNodeSelectorRequirementsWithMinValues(createdNodeClaim.Spec.Requirements...)
Expect(reqs).ToNot(HaveKey(v1.NodeRegisteredLabelKey))
Expect(reqs).ToNot(HaveKey(v1.NodeInitializedLabelKey))
Expect(createdNodeClaim.Labels).ToNot(HaveKey(v1.NodeRegisteredLabelKey))
Expect(createdNodeClaim.Labels).ToNot(HaveKey(v1.NodeInitializedLabelKey))
```

I verified locally that these assertions pass on your branch. They use existing patterns in the file (`cloudProvider.CreateCalls[0]` and `pscheduling.NewNodeSelectorRequirementsWithMinValues`) and would guard against a regression in the requirement-stripping logic in `ToNodeClaim()` and `resolveCustomLabelsFromRequirements()`.

1mwataru committed:
test: verify simulation-only requirements don't leak into NodeClaim

Add assertions to confirm that karpenter.sh/registered and
karpenter.sh/initialized requirements and labels are excluded
from the created NodeClaim, validating the exclusion logic in
ToNodeClaim() and resolveCustomLabelsFromRequirements().

Signed-off-by: Wataru Imai <imwata@amazon.com>

1mwataru commented:
@AndrewMitchell25 Thanks for the suggestion! 

I've incorporated your assertions to verify that simulation-only requirements don't leak into the created NodeClaim.

Regarding the CI failure — the `vulncheck` job is failing due to GO-2026-5026 (golang.org/x/net/idna), which is unrelated to this PR. 

I've opened #3041 3041to bump `golang.org/x/net` to v0.55.0 to resolve it.

Thank you

AndrewMitchell25 reviewed (approved):
LGTM

1mwataru commented:
Hi @jmdeal,

This PR has been reviewed and approved, and all review feedback is addressed. 
Could you take a look  when you get a chance?

Thank you!

jmdeal commented:
/lgtm
/approve
